Exported variable 'X' has or is using name 'Y' from external module "Z" but cannot be named.

**TypeScript Version:**  2.7.0-dev.20171212

**Code**

```ts
// thingA.ts
export interface ThingA { }

// thingB.ts
export interface ThingB { }

// things.ts
export { ThingA } from "./thingA";
export * from "./thingB";

// index.ts
import * as things from "./things";
export const thing1 = (param: things.ThingA) => null;
export const thing2 = (param: things.ThingB) => null;

// tsconfig.json
{
  "compilerOptions": {
    "declaration": true
  }
}
```

**Expected behavior:**

Project compiles without errors.

**Actual behavior:**

```
index.ts(3,14): error TS4023: Exported variable 'thing2' has or is using name 'ThingB' from external module "/Users/rob/Projects/tsctest/thingB" but cannot be named.
```

**Additional information:**

@weswigham fixed the following very similar issues in #19852:

* #5938
* #8612
* #19825

In fact, before his fix, line 2 of index.ts would also fail.

I've pushed this repro as a project here: https://github.com/robyoder/tsctest

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -2217,7 +2217,7 @@ namespace ts {
                         // Look in the exported members, if we can find accessibleSymbolChain, symbol is accessible using this chain
                         // but only if the symbolFromSymbolTable can be qualified
                         const candidateTable = getExportsOfSymbol(resolvedImportedSymbol);
-                        const accessibleSymbolsFromExports = candidateTable ? getAccessibleSymbolChainFromSymbolTable(candidateTable) : undefined;
+                        const accessibleSymbolsFromExports = candidateTable && getAccessibleSymbolChainFromSymbolTable(candidateTable);
                         if (accessibleSymbolsFromExports && canQualifySymbol(symbolFromSymbolTable, getQualifiedLeftMeaning(meaning))) {
                             return [symbolFromSymbolTable].concat(accessibleSymbolsFromExports);
                         }

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -2216,7 +2216,8 @@ namespace ts {
 
                         // Look in the exported members, if we can find accessibleSymbolChain, symbol is accessible using this chain
                         // but only if the symbolFromSymbolTable can be qualified
-                        const accessibleSymbolsFromExports = resolvedImportedSymbol.exports ? getAccessibleSymbolChainFromSymbolTable(resolvedImportedSymbol.exports) : undefined;
+                        const candidateTable = getExportsOfSymbol(resolvedImportedSymbol);
+                        const accessibleSymbolsFromExports = candidateTable ? getAccessibleSymbolChainFromSymbolTable(candidateTable) : undefined;
                         if (accessibleSymbolsFromExports && canQualifySymbol(symbolFromSymbolTable, getQualifiedLeftMeaning(meaning))) {
                             return [symbolFromSymbolTable].concat(accessibleSymbolsFromExports);
                         }

--- a/tests/baselines/reference/declarationEmitAliasExportStar.js
+++ b/tests/baselines/reference/declarationEmitAliasExportStar.js
@@ -1,21 +1,14 @@
 //// [tests/cases/compiler/declarationEmitAliasExportStar.ts] ////
 
-//// [thingA.ts]
-export interface ThingA { }
 //// [thingB.ts]
 export interface ThingB { }
 //// [things.ts]
-export { ThingA } from "./thingA";
 export * from "./thingB";
 //// [index.ts]
 import * as things from "./things";
-export const thing1 = (param: things.ThingA) => null;
 export const thing2 = (param: things.ThingB) => null;
 
 
-//// [thingA.js]
-"use strict";
-exports.__esModule = true;
 //// [thingB.js]
 "use strict";
 exports.__esModule = true;
@@ -25,20 +18,14 @@ exports.__esModule = true;
 //// [index.js]
 "use strict";
 exports.__esModule = true;
-exports.thing1 = function (param) { return null; };
 exports.thing2 = function (param) { return null; };
 
 
-//// [thingA.d.ts]
-export interface ThingA {
-}
 //// [thingB.d.ts]
 export interface ThingB {
 }
 //// [things.d.ts]
-export { ThingA } from "./thingA";
 export * from "./thingB";
 //// [index.d.ts]
 import * as things from "./things";
-export declare const thing1: (param: things.ThingA) => any;
 export declare const thing2: (param: things.ThingB) => any;

--- a/tests/baselines/reference/declarationEmitAliasExportStar.js
+++ b/tests/baselines/reference/declarationEmitAliasExportStar.js
@@ -1,0 +1,44 @@
+//// [tests/cases/compiler/declarationEmitAliasExportStar.ts] ////
+
+//// [thingA.ts]
+export interface ThingA { }
+//// [thingB.ts]
+export interface ThingB { }
+//// [things.ts]
+export { ThingA } from "./thingA";
+export * from "./thingB";
+//// [index.ts]
+import * as things from "./things";
+export const thing1 = (param: things.ThingA) => null;
+export const thing2 = (param: things.ThingB) => null;
+
+
+//// [thingA.js]
+"use strict";
+exports.__esModule = true;
+//// [thingB.js]
+"use strict";
+exports.__esModule = true;
+//// [things.js]
+"use strict";
+exports.__esModule = true;
+//// [index.js]
+"use strict";
+exports.__esModule = true;
+exports.thing1 = function (param) { return null; };
+exports.thing2 = function (param) { return null; };
+
+
+//// [thingA.d.ts]
+export interface ThingA {
+}
+//// [thingB.d.ts]
+export interface ThingB {
+}
+//// [things.d.ts]
+export { ThingA } from "./thingA";
+export * from "./thingB";
+//// [index.d.ts]
+import * as things from "./things";
+export declare const thing1: (param: things.ThingA) => any;
+export declare const thing2: (param: things.ThingB) => any;

--- a/tests/baselines/reference/declarationEmitAliasExportStar.symbols
+++ b/tests/baselines/reference/declarationEmitAliasExportStar.symbols
@@ -1,29 +1,16 @@
-=== tests/cases/compiler/thingA.ts ===
-export interface ThingA { }
->ThingA : Symbol(ThingA, Decl(thingA.ts, 0, 0))
-
 === tests/cases/compiler/thingB.ts ===
 export interface ThingB { }
 >ThingB : Symbol(ThingB, Decl(thingB.ts, 0, 0))
 
 === tests/cases/compiler/things.ts ===
-export { ThingA } from "./thingA";
->ThingA : Symbol(ThingA, Decl(things.ts, 0, 8))
-
 export * from "./thingB";
-=== tests/cases/compiler/index.ts ===
+No type information for this code.=== tests/cases/compiler/index.ts ===
 import * as things from "./things";
 >things : Symbol(things, Decl(index.ts, 0, 6))
 
-export const thing1 = (param: things.ThingA) => null;
->thing1 : Symbol(thing1, Decl(index.ts, 1, 12))
->param : Symbol(param, Decl(index.ts, 1, 23))
->things : Symbol(things, Decl(index.ts, 0, 6))
->ThingA : Symbol(things.ThingA, Decl(things.ts, 0, 8))
-
 export const thing2 = (param: things.ThingB) => null;
->thing2 : Symbol(thing2, Decl(index.ts, 2, 12))
->param : Symbol(param, Decl(index.ts, 2, 23))
+>thing2 : Symbol(thing2, Decl(index.ts, 1, 12))
+>param : Symbol(param, Decl(index.ts, 1, 23))
 >things : Symbol(things, Decl(index.ts, 0, 6))
 >ThingB : Symbol(things.ThingB, Decl(thingB.ts, 0, 0))
 

--- a/tests/baselines/reference/declarationEmitAliasExportStar.symbols
+++ b/tests/baselines/reference/declarationEmitAliasExportStar.symbols
@@ -1,0 +1,29 @@
+=== tests/cases/compiler/thingA.ts ===
+export interface ThingA { }
+>ThingA : Symbol(ThingA, Decl(thingA.ts, 0, 0))
+
+=== tests/cases/compiler/thingB.ts ===
+export interface ThingB { }
+>ThingB : Symbol(ThingB, Decl(thingB.ts, 0, 0))
+
+=== tests/cases/compiler/things.ts ===
+export { ThingA } from "./thingA";
+>ThingA : Symbol(ThingA, Decl(things.ts, 0, 8))
+
+export * from "./thingB";
+=== tests/cases/compiler/index.ts ===
+import * as things from "./things";
+>things : Symbol(things, Decl(index.ts, 0, 6))
+
+export const thing1 = (param: things.ThingA) => null;
+>thing1 : Symbol(thing1, Decl(index.ts, 1, 12))
+>param : Symbol(param, Decl(index.ts, 1, 23))
+>things : Symbol(things, Decl(index.ts, 0, 6))
+>ThingA : Symbol(things.ThingA, Decl(things.ts, 0, 8))
+
+export const thing2 = (param: things.ThingB) => null;
+>thing2 : Symbol(thing2, Decl(index.ts, 2, 12))
+>param : Symbol(param, Decl(index.ts, 2, 23))
+>things : Symbol(things, Decl(index.ts, 0, 6))
+>ThingB : Symbol(things.ThingB, Decl(thingB.ts, 0, 0))
+

--- a/tests/baselines/reference/declarationEmitAliasExportStar.types
+++ b/tests/baselines/reference/declarationEmitAliasExportStar.types
@@ -1,27 +1,12 @@
-=== tests/cases/compiler/thingA.ts ===
-export interface ThingA { }
->ThingA : ThingA
-
 === tests/cases/compiler/thingB.ts ===
 export interface ThingB { }
 >ThingB : ThingB
 
 === tests/cases/compiler/things.ts ===
-export { ThingA } from "./thingA";
->ThingA : any
-
 export * from "./thingB";
-=== tests/cases/compiler/index.ts ===
+No type information for this code.=== tests/cases/compiler/index.ts ===
 import * as things from "./things";
 >things : typeof things
-
-export const thing1 = (param: things.ThingA) => null;
->thing1 : (param: things.ThingA) => any
->(param: things.ThingA) => null : (param: things.ThingA) => any
->param : things.ThingA
->things : any
->ThingA : things.ThingA
->null : null
 
 export const thing2 = (param: things.ThingB) => null;
 >thing2 : (param: things.ThingB) => any

--- a/tests/baselines/reference/declarationEmitAliasExportStar.types
+++ b/tests/baselines/reference/declarationEmitAliasExportStar.types
@@ -1,0 +1,33 @@
+=== tests/cases/compiler/thingA.ts ===
+export interface ThingA { }
+>ThingA : ThingA
+
+=== tests/cases/compiler/thingB.ts ===
+export interface ThingB { }
+>ThingB : ThingB
+
+=== tests/cases/compiler/things.ts ===
+export { ThingA } from "./thingA";
+>ThingA : any
+
+export * from "./thingB";
+=== tests/cases/compiler/index.ts ===
+import * as things from "./things";
+>things : typeof things
+
+export const thing1 = (param: things.ThingA) => null;
+>thing1 : (param: things.ThingA) => any
+>(param: things.ThingA) => null : (param: things.ThingA) => any
+>param : things.ThingA
+>things : any
+>ThingA : things.ThingA
+>null : null
+
+export const thing2 = (param: things.ThingB) => null;
+>thing2 : (param: things.ThingB) => any
+>(param: things.ThingB) => null : (param: things.ThingB) => any
+>param : things.ThingB
+>things : any
+>ThingB : things.ThingB
+>null : null
+

--- a/tests/cases/compiler/declarationEmitAliasExportStar.ts
+++ b/tests/cases/compiler/declarationEmitAliasExportStar.ts
@@ -1,12 +1,8 @@
 // @declaration: true
-// @filename: thingA.ts
-export interface ThingA { }
 // @filename: thingB.ts
 export interface ThingB { }
 // @filename: things.ts
-export { ThingA } from "./thingA";
 export * from "./thingB";
 // @filename: index.ts
 import * as things from "./things";
-export const thing1 = (param: things.ThingA) => null;
 export const thing2 = (param: things.ThingB) => null;

--- a/tests/cases/compiler/declarationEmitAliasExportStar.ts
+++ b/tests/cases/compiler/declarationEmitAliasExportStar.ts
@@ -1,0 +1,12 @@
+// @declaration: true
+// @filename: thingA.ts
+export interface ThingA { }
+// @filename: thingB.ts
+export interface ThingB { }
+// @filename: things.ts
+export { ThingA } from "./thingA";
+export * from "./thingB";
+// @filename: index.ts
+import * as things from "./things";
+export const thing1 = (param: things.ThingA) => null;
+export const thing2 = (param: things.ThingB) => null;

